iwyu_include_picker.cc

```diff
@@ -1330,18 +1330,50 @@ string IncludePicker::MaybeGetIncludeNameAsWritten(
   return value ? *value : "";
 }
 
-vector<string> IncludePicker::GetCandidateHeadersForSymbol(
-    const string& symbol) const {
-  CHECK_(has_called_finalize_added_include_lines_ && "Must finalize includes");
-  vector<MappedInclude> mapped_includes =
-    GetPublicValues(symbol_include_map_, symbol);
+vector<string> IncludePicker::BestQuotedIncludesForIncluder(
+    const vector<MappedInclude>& includes,
+    const string& including_filepath) const {
+  // Convert each MappedInclude to a quoted include.  Each contains
+  // a default quoted include but sometimes we can do better.
+  // For each path, the iwyu-preprocessor may have stored the quoted-include
+  // as written in including_filepath.  This is better to use than
+  // ConvertToQuotedInclude because it avoids trouble when the same
+  // file is accessible via different include search-paths, or is
+  // accessed via a symlink.
+  // We also try to recalculate any quoted includes using absolute paths,
+  // because they can likely be converted to an include relative to the
+  // current file.
+  const string including_path =
+      MakeAbsolutePath(GetParentPath(including_filepath));
   vector<string> retval;
-  for (const MappedInclude& mapped_include : mapped_includes) {
-    retval.push_back(mapped_include.quoted_include);
+  for (const MappedInclude& mapped_include : includes) {
```

Comment:
Could you add one or more comments to describe the priority order for "best" now? It seems like the quoted-include as written is the least favorite, which comes as a bit of a surprise to me.

I guess that's what the mega-comment above explains, but I wonder if it would be easier to take in as a numbered list or something.

Reply:
Overall, this commit is really nice and focused now, but I can't for the life of me connect the dots to make me understand why it works. :)

Reply:
I've rewritten the comment.

The as-written version is still the top priority; that's why it's the first alternative in the `if` statement.